### PR TITLE
attempt to optimize energy grid ticking

### PR DESCRIPTION
**Pull request by Glease**
1. early exit for power consumption call if a creative cell is in any networks connected to it via quartz fiber
2. added a single slot LRU cache for delegated power extraction, so deeply nested subnets will always try to pick up power from the main grid, without the possiblity of recursing into other subnets. This cache will partially fail if player place his power cells in more than one subnet, and need more than one of them to buffer the energy consumption, but that is rather rare.

### CODE REVIEW

**Comment by firenoo** [comment]
I think there are some missed opportunities for more aggressive optimizations. Currently, you'll eventually run into the same performance issues using quartz fiber (it does not even have to be nested too far deep).
I suppose for *most* applications it will net a gain.

If a grid has a creative source such as a creative energy cell attached, the grid itself can be considered a creative source, and skip energy computations. This may cause additional computation when updating the grid, but I think that the penalty is outweighed by the performance savings in a short time.

**Comment by Glease** [comment]
@firenoo I'm not totally sure what you are referring to by "the grid itself can be considered a creative source, and skip energy computations". In my current implementation, whenever a node is added/remove or net merged/split all connected me components will be scanned for creative cells. If any is found all EnergyGridCache involved will has the infinite flag set

**Comment by firenoo** [comment]
Ah, I found out what was muddying my tests. I was using this setup:

image

Basically, what I did was I put a breakpoint to see if a Grid was infinite. But the breakpoint wasn't being hit 100% of the time, so I was wondering what was causing the Grid to not be infinite. As it turns out, there is a fake Grid created by the testing framework that I did not know existed, which was not infinite was what was causing the disparity. Sorry about that

**Comment by firenoo** [comment]
Is this ready?

**Comment by Glease** [comment]
smh this got lost on my personal youtrack... yeah this is ready, but it has only received some minimal testings on sampas' server, and he reports that there are still significant tick time usage over energy cache: https://discord.com/channels/181078474394566657/939305179524792340/1137384509126479953

File: src/main/java/appeng/me/cache/EnergyGridCache.java [hunk 1]
```diff
@@ -262,6 +275,31 @@ public double extractAEPower(final double amt, final Actionable mode, final Set<
         return extractedPower;
     }
 
+    private double extractFromOtherGrids(double amt, Actionable mode, Set<IEnergyGrid> seen, double extractedPower) {
+        IEnergyGridProvider energyGridProvider = lastGridProvider.get();
+        if (energyGridProvider != null) {
```
Comment (Laiff) [inline src/main/java/appeng/me/cache/EnergyGridCache.java:280]: Added ability to cache last direction to extract power

File: src/main/java/appeng/me/cache/EnergyGridCache.java [hunk 2]
```diff
@@ -433,10 +492,23 @@ public double getEnergyDemand(final double maxRequired) {
...
+        for (IEnergyGrid grid : grids) {
+            grid.setHasInfiniteStore(infinite);
+        }
+    }
+
     @Override
     public void removeNode(final IGridNode node, final IGridHost machine) {
         if (machine instanceof IEnergyGridProvider) {
             this.energyGridProviders.remove(machine);
+            if (lastGridProvider.get() == machine) {
```
Comment (Laiff) [inline src/main/java/appeng/me/cache/EnergyGridCache.java:507]: Cached provider should be cleared to correct update grid integrity

File: src/main/java/appeng/parts/networking/PartQuartzFiber.java [hunk 3]
```diff
@@ -46,6 +47,8 @@ public class PartQuartzFiber extends AEBasePart implements IEnergyGridProvider {
             this.getProxy().getMachineRepresentation(),
             true);
 
+    private WeakReference<IEnergyGrid> lastEnergyGrid = new WeakReference<>(null);
```
Comment (Laiff) [inline src/main/java/appeng/parts/networking/PartQuartzFiber.java:50]: cache for last energy grid with with positive energy balance

**Review by Laiff — APPROVED** [review]

**Pull request merged**